File: js/src/resources/isolate-audio.ts
import type { HttpClient, PollingOptions, RequestOptions } from '@runapi.ai/core';
import { compactParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { IsolateAudioParams, AudioTaskResponse, CompletedAudioTaskResponse, TaskCreateResponse } from '../types';

const ENDPOINT = '/api/v1/elevenlabs/isolate_audio';

export class IsolateAudio {
  constructor(private readonly http: HttpClient) {}

  async run(params: IsolateAudioParams, options?: RequestOptions & PollingOptions): Promise<CompletedAudioTaskResponse> {
    const { id } = await this.create(params, options);
    const response = await pollUntilComplete<AudioTaskResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
    return response as CompletedAudioTaskResponse;
  }

  async create(params: IsolateAudioParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body: compactParams(params),
      ...options,
    });
  }

  async get(id: string, options?: RequestOptions): Promise<AudioTaskResponse> {
    return this.http.request<AudioTaskResponse>('GET', `${ENDPOINT}/${id}`, options ?? {});
  }
}

File: ruby/lib/runapi/elevenlabs/resources/isolate_audio.rb
# frozen_string_literal: true

module RunApi
  module Elevenlabs
    module Resources
      class IsolateAudio
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/elevenlabs/isolate_audio"
        RESPONSE_CLASS = Types::AudioTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedAudioTaskResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          raise Core::ValidationError, "audio_url is required" unless param(params, :audio_url)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end
      end
    end
  end
end
